src/systems/buoyancy/Buoyancy.cc

```diff
@@ -243,25 +395,76 @@ void Buoyancy::PreUpdate(const ignition::gazebo::UpdateInfo &_info,
       // World pose of the link.
       math::Pose3d linkWorldPose = worldPose(_entity, _ecm);
 
+      Link link(_entity);
+
+      math::Vector3d buoyancy;
       // By Archimedes' principle,
       // buoyancy = -(mass*gravity)*fluid_density/object_density
       // object_density = mass/volume, so the mass term cancels.
-      math::Vector3d buoyancy =
-        -this->dataPtr->FluidDensity(linkWorldPose) *
+      if(this->dataPtr->buoyancyType
+        == BuoyancyPrivate::BuoyancyType::UNIFORM_BUOYANCY)
+      {
+        buoyancy =
+        -this->dataPtr->UniformFluidDensity(linkWorldPose) *
         _volume->Data() * gravity->Data();
 
-      // Convert the center of volume to the world frame
-      math::Vector3d offsetWorld = linkWorldPose.Rot().RotateVector(
-          _centerOfVolume->Data());
-      // Compute the torque that should be applied due to buoyancy and
-      // the center of volume.
-      math::Vector3d torque = offsetWorld.Cross(buoyancy);
+        // Convert the center of volume to the world frame
+        math::Vector3d offsetWorld = linkWorldPose.Rot().RotateVector(
+            _centerOfVolume->Data());
+        // Compute the torque that should be applied due to buoyancy and
+        // the center of volume.
+        math::Vector3d torque = offsetWorld.Cross(buoyancy);
+
+        // Apply the wrench to the link. This wrench is applied in the
+        // Physics System.
+        link.AddWorldWrench(_ecm, buoyancy, torque);
+      }
+      else if(this->dataPtr->buoyancyType
+        == BuoyancyPrivate::BuoyancyType::GRADED_BUOYANCY)
+      {
+        std::vector<Entity> collisions = _ecm.ChildrenByComponents(
+          _entity, components::Collision());
+        this->dataPtr->buoyancyForces.clear();
+
+        for(auto e : collisions)
+        {
+          const components::CollisionElement *coll =
+            _ecm.Component<components::CollisionElement>(e);
 
+          auto pose = worldPose(e, _ecm);
+
+          if (!coll)
+          {
+            ignerr << "Invalid collision pointer. This shouldn't happen\n";
+            continue;
+          }
+
+          switch (coll->Data().Geom()->Type())
+          {
+            case sdf::GeometryType::BOX:
+              this->dataPtr->GradedFluidDensity<math::Boxd>(
+                pose,
+                coll->Data().Geom()->BoxShape()->Shape(),
+                gravity->Data());
+              break;
+            case sdf::GeometryType::SPHERE:
+              this->dataPtr->GradedFluidDensity<math::Sphered>(
+                pose,
+                coll->Data().Geom()->SphereShape()->Shape(),
+                gravity->Data());
+              break;
+            default:
+              ignerr << "Only <box> and <sphere> collisions are supported by "
+                << "the graded buoyancy option." << std::endl;
```

Comment:
How about making this a warning? The user may have other shapes in the world which they don't expect to float.